src/process-events.c

```diff
@@ -116,34 +167,49 @@ static __always_inline void* is_user_process(void *ts)
 }
 
 // pushes a message to the process_events perfmap for the current
-// CPU. The message is purposefully not cleared so its data (in
-// particular ev->event_id, and ev->type) can be reused if
-// necessary.
-static __always_inline void push_message(struct pt_regs *ctx, pprocess_message_t ev)
-{
-    bpf_perf_event_output(ctx, &process_events, BPF_F_CURRENT_CPU, ev, sizeof(*ev));
-}
-
-// sends PM_ENTER_DONE to signal the end of an enter_* event (i.e, a
-// kprobe). Re-uses ev->event_id but sets the other fields.
-static __always_inline void push_enter_done(struct pt_regs *ctx, pprocess_message_t ev)
-{
-    ev->type = PM_ENTER_DONE;
+// CPU.
+static __always_inline int push_message(struct pt_regs *ctx, pprocess_message_t ev) {
+    return bpf_perf_event_output(ctx, &process_events, BPF_F_CURRENT_CPU, ev, sizeof(*ev));
+}
+
+// pushes a message with an extra `dynamic_size` number of bytes. It
+// caps the message to `MAX_PERCPU_BUFFER - 1` to appease verifier
+// however. The "-1" is there to make it an easy bit cap against a
+// power of two; it may drop an null byte out of the string which is
+// OK.
+static __always_inline int push_flexible_message(struct pt_regs *ctx, pprocess_message_t ev, u64 dynamic_size) {
+    return bpf_perf_event_output(ctx, &process_events, BPF_F_CURRENT_CPU, ev, dynamic_size & (MAX_PERCPU_BUFFER - 1));
+}
+
+// writes the string into the provided buffer w/ offset. On a
+// succesful write it modifies the offset with the length of the read
+// string. It deliberately does not handle truncations; it just reads
+// up to `max_string`.
+static __always_inline int write_string(const char *string, buf_t *buffer, u32 *offset, const u32 max_string) {
+    // A smarter implementation of this wouldn't use max_string but
+    // instead would just check MAX_PERCPU_BUFFER - *offset as the max
+    // that it can write. However, the verifier seems allergic to an
+    // smarter implementation as it complains about potential out of
+    // bounds or negative values. While perhaps theoretically possible
+    // to improve this with just the right incantations (and maybe
+    // turning off some compiler optimizaitons that remove some
+    // checks) at this time this is considered good enough (TM).
+
+    // already too full
+    if (*offset > MAX_PERCPU_BUFFER - max_string) return 0;
+
+    int sz = bpf_probe_read_str(&buffer->buf[*offset], max_string, string);
```

Comment:
This is the part I recall the verifier having the most trouble with. If you can get it to work, this is a much better way to send information for all the reasons you already listed.

Reply:
Yeah that part gave me a LOT of issues when I was trying to be smart about it (hence the giant comment right above it). Now that the "max_string" is hardcoded then the verifier is happy as long as the bounds check is very close to it (the if statement right above it). I spent a good chunk of time trying to make it work with fancier code, even adding compiler barrier (asm volatile) and it just wouldn't like it so I gave up and made it dumb enough for the verifier to be happy with it

Reply:
Yep, that's the verifier.

Reply:
I have tested that it works on:

* Ubuntu 16.04 4.15.0-112-generic
* Ubuntu 18.04 5.4.0-113-generic
* Centos 7 3.10.0-1160.53.1.el7.x86_64
* Centos 8 4.18.0-365.el8.x86_64